### PR TITLE
TL/UCP: reorder ranks for SRA

### PR DESCRIPTION
**Pull request by shimmybalsam**
## What
Add reorder ranks in for Allreduce SRA

## Why ?
Performance benfit, especialy when n_extra ranks > 0, in which case we do a complete allreduce between "proxy" and "extra" ranks and if they are on the same node this significantly improves perf.

## How ?
By reordring ranks in each step:  reduce_scatter_knomial, allgather_knomial.

**Note:** allgather_knomial is used also for small message allgather and large message bcast (bcast SAG). In these cases reorder ranks will not be used, it is currently only implemented for the usage within allreduce SRA.

### CODE REVIEW

**Comment by shimmybalsam** [comment]
**Without extra ranks:**

image


**With extra ranks:**

image

**Comment by shimmybalsam** [comment]
bot:retest

File: src/components/tl/ucp/allgather/allgather_knomial.c [hunk 1]
```diff
@@ -230,14 +238,22 @@ ucc_status_t ucc_tl_ucp_allgather_knomial_init_r(
     ucc_base_coll_args_t *coll_args, ucc_base_team_t *team,
     ucc_coll_task_t **task_h, ucc_kn_radix_t radix)
 {
+    ucc_tl_ucp_team_t *tl_team = ucc_derived_of(team, ucc_tl_ucp_team_t);
     ucc_tl_ucp_task_t *task;
+    ucc_sbgp_t        *sbgp;
 
     task = ucc_tl_ucp_init_task(coll_args, team);
+    if (tl_team->cfg.use_reordering &&
+        coll_args->args.coll_type == UCC_COLL_TYPE_ALLREDUCE) {
```
Comment (samnordmann) [inline src/components/tl/ucp/allgather/allgather_knomial.c:247]: Can you explain why we only enable it for ALLREDUCE?

**Review by samnordmann — APPROVED** [review]

File: src/components/tl/ucp/allgather/allgather_knomial.c (hunk 1 above)
Comment (shimmybalsam) [reply src/components/tl/ucp/allgather/allgather_knomial.c:247]: allgather_knomial is used also for small message allgather and large message bcast (bcast SAG). In these cases reorder ranks will not be used yet - additional changes will need to be made.
Currently in this PR, the only usage of reorder ranks in allgather that is implemented is as part of allreduce SRA.

**Review by Sergei-Lebedev — COMMENTED** [review]
can we instead make map part of knomial pattern. That way we don't need to repeat ucc_ep_map_eval every time we get peer.
In future we can also add ring patter to do the same for ring algorithms

**Comment by shimmybalsam** [comment]
> can we instead make map part of knomial pattern. That way we don't need to repeat ucc_ep_map_eval every time we get peer. In future we can also add ring patter to do the same for ring algorithms

As discussed on call, in my opinion we should merge this PR as is as first step, since it substantially improves performance of SRA. In future we can rewrite knomial pattern functions (which will effect all knomial algorithms in TL/UCP).

Meanwhile here is some more supportive data:

image

**Comment by shimmybalsam** [comment]
bot:retest

**Review by Sergei-Lebedev — APPROVED** [review]

**Pull request merged**